File: sciurus17_examples/src/neck_control.cpp
// Copyright 2023 RT Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Reference:
// https://github.com/ros-planning/moveit2_tutorials/blob/humble/doc/
// examples/move_group_interface/src/move_group_interface_tutorial.cpp

#include <memory>
#include <string>
#include <thread>
#include <vector>

#include <angles/angles.h>
#include <moveit/move_group_interface/move_group_interface.hpp>
#include <rclcpp/rclcpp.hpp>

using MoveGroupInterface = moveit::planning_interface::MoveGroupInterface;

class NeckControl
{
public:
  // ノードを受け取り、首のMoveGroupInterfaceを初期化する
  explicit NeckControl(rclcpp::Node::SharedPtr node)
  {
    move_group_neck_ = std::make_shared<MoveGroupInterface>(node, "neck_group");
    move_group_neck_->setMaxVelocityScalingFactor(0.1);  // Set 0.0 ~ 1.0
    move_group_neck_->setMaxAccelerationScalingFactor(0.1);  // Set 0.0 ~ 1.0
  }

  // SRDFに定義された姿勢名で首を動かす
  void move_to_named_pose(const std::string & name)
  {
    move_group_neck_->setNamedTarget(name);
    move_group_neck_->move();
  }

  // 各ジョイント角度[rad]を指定して首を動かす
  void move_joint_values(const std::vector<double> & joint_values)
  {
    move_group_neck_->setJointValueTarget(joint_values);
    move_group_neck_->move();
  }

  // 首の現在のジョイント角度を取得する
  std::vector<double> get_current_joint_values()
  {
    return move_group_neck_->getCurrentJointValues();
  }

private:
  std::shared_ptr<MoveGroupInterface> move_group_neck_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::NodeOptions node_options;
  node_options.automatically_declare_parameters_from_overrides(true);
  auto node = rclcpp::Node::make_shared("neck_control", node_options);

  // MoveGroupInterfaceのデッドロックを防ぐため、スピン処理を別スレッドで走らせる
  std::thread spin_thread([node]() {rclcpp::spin(node);});

  NeckControl controller(node);

  // SRDFに定義されている"neck_init_pose"の姿勢にする
  controller.move_to_named_pose("neck_init_pose");

  // 現在角度をベースに、目標角度を作成する
  auto joint_values = controller.get_current_joint_values();

  // 首を左に向ける
  joint_values[0] = angles::from_degrees(45.0);
  controller.move_joint_values(joint_values);

  // 首を右に向ける
  joint_values[0] = angles::from_degrees(-45.0);
  controller.move_joint_values(joint_values);

  // 首を前に向ける
  joint_values[0] = angles::from_degrees(0.0);
  controller.move_joint_values(joint_values);

  // 首を上に向ける
  joint_values[1] = angles::from_degrees(45.0);
  controller.move_joint_values(joint_values);

  // 首を下に向ける
  joint_values[1] = angles::from_degrees(-45.0);
  controller.move_joint_values(joint_values);

  // "neck_init_pose"に戻す
  controller.move_to_named_pose("neck_init_pose");

  rclcpp::shutdown();
  spin_thread.join();
  return 0;
}

File: sciurus17_examples_py/sciurus17_examples_py/neck_control.py
# Copyright 2024 RT Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math

from moveit.core.robot_state import RobotState
from moveit.planning import (
    MoveItPy,
    PlanRequestParameters,
)

import rclpy
from rclpy.logging import get_logger

from sciurus17_examples_py.utils import plan_and_execute


class NeckControl:
    def __init__(self):
        # MoveItPyのインスタンスを生成し、planning componentを取得
        self.sciurus17 = MoveItPy(node_name='neck_control')
        self.logger = get_logger('neck_control')

        # 首制御用 planning component
        self.neck = self.sciurus17.get_planning_component('neck_group')

        # planning scene monitor（現在のジョイント角度取得に使用）
        self.planning_scene_monitor = self.sciurus17.get_planning_scene_monitor()

        # ロボットモデルの取得（ジョイント目標値の設定に使用）
        self.robot_model = self.sciurus17.get_robot_model()

        # プランニングの設定（動作プランナーと速度・加速度スケール）
        self.neck_plan_params = PlanRequestParameters(self.sciurus17, 'ompl_rrtc_default')
        self.neck_plan_params.max_velocity_scaling_factor = 0.1  # Set 0.0 ~ 1.0
        self.neck_plan_params.max_acceleration_scaling_factor = 0.1  # Set 0.0 ~ 1.0

    def move_to_named_pose(self, configuration_name):
        # SRDFに定義された姿勢名で首を動かす
        self.neck.set_start_state_to_current_state()
        self.neck.set_goal_state(configuration_name=configuration_name)
        plan_and_execute(
            self.sciurus17, self.neck, self.logger,
            single_plan_parameters=self.neck_plan_params,
        )

    def move_joint_values(self, joint_values):
        # ジョイント角度[rad]のリストを指定して首を動かす
        self.neck.set_start_state_to_current_state()
        robot_state = RobotState(self.robot_model)
        robot_state.set_joint_group_positions('neck_group', joint_values)
        self.neck.set_goal_state(robot_state=robot_state)
        plan_and_execute(
            self.sciurus17, self.neck, self.logger,
            single_plan_parameters=self.neck_plan_params,
        )

    def get_current_joint_values(self):
        # 首の現在のジョイント角度をリスト形式で取得する
        joint_values = []
        with self.planning_scene_monitor.read_only() as scene:
            robot_state = scene.current_state
            joint_values = robot_state.get_joint_group_positions('neck_group')
        return joint_values


def main(args=None):
    rclpy.init(args=args)

    controller = NeckControl()

    # 初期姿勢に移動
    controller.move_to_named_pose('neck_init_pose')

    # 現在の首のジョイント角度を取得
    joint_values = controller.get_current_joint_values()

    # 首を左に向ける
    joint_values[0] = math.radians(45.0)
    controller.move_joint_values(joint_values)

    # 首を右に向ける
    joint_values[0] = math.radians(-45.0)
    controller.move_joint_values(joint_values)

    # 首を前に向ける
    joint_values[0] = math.radians(0.0)
    controller.move_joint_values(joint_values)

    # 首を上に向ける
    joint_values[1] = math.radians(45.0)
    controller.move_joint_values(joint_values)

    # 首を下に向ける
    joint_values[1] = math.radians(-45.0)
    controller.move_joint_values(joint_values)

    # 初期姿勢に戻す
    controller.move_to_named_pose('neck_init_pose')

    # Finish with error. Related Issue
    # https://github.com/moveit/moveit2/issues/2693
    rclpy.shutdown()


if __name__ == '__main__':
    main()
